coordinator/coordinator.go

```diff
@@ -35,34 +36,81 @@ func (c *Coordinator) Run() error {
+	// The value of users to be decremented at each iteration.
+	// It should be proportional to the maximum number of users expected to test.
+	const decValue = 8
+	// The timespan to wait after a performance degradation alert before
+	// incrementing or decrementing users again.
+	const restTime = 10 * time.Second
 
 	for {
-		status := c.cluster.Status()
-		mlog.Info("coordinator: cluster status:", mlog.Int("active_users", status.ActiveUsers), mlog.Int64("errors", status.NumErrors))
-
-		if status.ActiveUsers < c.config.ClusterConfig.MaxActiveUsers {
-			// TODO: make the choice of this value a bit smarter.
-			inc := 10
-			diff := c.config.ClusterConfig.MaxActiveUsers - status.ActiveUsers
-			if diff < inc {
-				inc = diff
-			}
-			mlog.Info("coordinator: incrementing active users", mlog.Int("num_users", inc))
-			err := c.cluster.IncrementUsers(inc)
-			if err != nil {
-				mlog.Error("coordinator: failed to increment users", mlog.Err(err))
-			}
-		}
+		var perfStatus performance.Status
 
 		select {
 		case <-interruptChannel:
 			mlog.Info("coordinator: shutting down")
-			c.cluster.Shutdown()
 			return nil
-		case <-time.After(1 * time.Second):
+		case perfStatus = <-monitorChan:
+		}
+
+		if perfStatus.Alert {
+			lastAlertT = time.Now()
 		}
 
-		// TODO: implement performance monitoring and act on them to complete feedback loop.
+		status := c.cluster.Status()
+		mlog.Debug("coordinator: cluster status:", mlog.Int("active_users", status.ActiveUsers), mlog.Int64("errors", status.NumErrors))
+
+		// supportedUsers should be estimated in a more clever way in the future.
+		// For now we say that the supported number of users is the number of active users that ran
+		// for the defined timespan without causing any performance degradation alert.
+		if !lastAlertT.IsZero() && !perfStatus.Alert && hasPassed(lastAlertT, restTime) && hasPassed(lastActionT, restTime) {
+			supportedUsers = status.ActiveUsers
+		}
+
+		mlog.Debug("coordinator: supported users", mlog.Int("supported_users", supportedUsers))
+
+		// We give the feedback loop some rest time in case of performance
+		// degradation alerts. We want metrics to stabilize before incrementing/decrementing users again.
+		if lastAlertT.IsZero() || lastActionT.IsZero() || hasPassed(lastActionT, restTime) {
+			if perfStatus.Alert {
+				if err := c.cluster.DecrementUsers(decValue); err != nil {
+					mlog.Error("coordinator: failed to decrement users", mlog.Err(err))
+				}
+				lastActionT = time.Now()
+			} else if lastAlertT.IsZero() || hasPassed(lastAlertT, restTime) {
+				if status.ActiveUsers < c.config.ClusterConfig.MaxActiveUsers {
+					inc := min(incValue, c.config.ClusterConfig.MaxActiveUsers-status.ActiveUsers)
+					mlog.Info("coordinator: incrementing active users", mlog.Int("num_users", inc))
+					if err := c.cluster.IncrementUsers(inc); err != nil {
+						mlog.Error("coordinator: failed to increment users", mlog.Err(err))
+					}
+					lastActionT = time.Now()
```

Comment:
I see that we are updating lastActionT irrespective of whether Increment/Decrement failed or not. What are the implications of IncrementUsers failing and lastActionT still getting updated ? Will there be any issues with it getting updated but we fail to reduce users, and now it is within the restTime and we don't take any action until then ?

Reply:
Worst case scenario we wait `restTime` before attempting to run the action again. 
I don't see any problem setting it only in case of a non error response though.